src/util/SweptCollection.js

```diff
@@ -0,0 +1,138 @@
+     * @type {number}
+     */
+    this.maxSize = maxSize;
+
+    /**
+     * A function called to exclude items from being removed when the Collection reaches max size.
+     * <warn>When the function returns `true` for every entry, the first entry is removed to maintain size</warn>
+     * @type {?Function}
+     */
+    this.keepAtMaxSize = keepAtMaxSize;
+
+    /**
+     * A function called every sweep interval that returns a function passed to `sweep`
+     * @type {?Function}
+     */
+    this.sweepFilter = sweepFilter;
+
+    /**
+     * The id of the interval being used to sweep.
+     * @type {?Number}
+     */
+    this.interval =
+      sweepInterval > 0 && sweepFilter ? setInterval(() => this.sweep(this.sweepFilter()), sweepInterval * 1000) : null;
+  }
+
+  set(key, value) {
+    if (this.maxSize === 0) return this;
+    if (this.maxSize === -1 || this.maxSize === Infinity) super.set(key, value);
+    if (this.size >= this.maxSize && !this.has(key)) {
+      let deleted = false;
+      if (this.keepAtMaxSize) {
+        for (const [k, v] of this) {
+          if (!this.keepAtMaxSize(v)) {
+            this.delete(k);
+            deleted = true;
+            break;
+          }
+        }
+      }
+      if (!deleted) {
+        this.delete(this.firstKey());
+      }
+    }
+    return super.set(key, value);
+  }
+
+  /**
+   * Options for generating a filter function based on lifetime
+   * @typedef {Object} LifetimeFilterOptions
+   * @property {number} [lifetime=14400] How long an entry should stay in the collection
+   * before it is considered sweepable
+   * @property {Function} [getComparisonTimestamp=`e => e.createdTimestamp`] A function that takes an entry
+   * and returns a timestamp to compare against in order to determine the lifetime of the entry.
+   * @property {Function} [excludeFromSweep=null] A function that takes an entry and returns a boolean,
+   * `true` when the entry should not be checked for sweepability.
+   */
+
+  /**
+   * Create a sweepFilter function that uses a lifetime to determine sweepability.
+   * @param {LifetimeFilterOptions} [options={}] The options used to generate the filter function
+   * @returns {Function}
+   */
+  static filterByLiftetme({
+    lifetime = 14400,
+    getComparisonTimestamp = e => e?.createdTimestamp,
+    excludeFromSweep = null,
+  } = {}) {
+    if (typeof lifetime !== 'number') throw new TypeError('INVALID_TYPE', 'lifetime', 'number');
+    if (typeof getComparisonTimestamp !== 'function') {
+      throw new TypeError('INVALID_TYPE', 'getComparisonTimestamp', 'function');
+    }
+    if (excludeFromSweep !== null && typeof excludeFromSweep !== 'function') {
+      throw new TypeError('INVALID_TYPE', 'excludeFromSweep', 'function');
+    }
+    return () => {
+      if (lifetime <= 0) return () => false;
+      const lifetimeMs = lifetime * 1000;
+      const now = Date.now();
+      return entry => {
+        if (excludeFromSweep?.(entry)) {
```

Comment:
Maybe make a special callback that always returns `false` when `excludeFromSweep` is null?

```javascript
if (excludeFromSweep === null) return () => false;
```

That way we can optimize this kind of operation, and we skip a check in this callback.

Reply:
You can also add this to line 123 with an `||` operator.

Reply:
This is the exact opposite of what its supposed to do, when null it should sweep everything. I've just changed the default to be `() => false` since it'll make it more clear